parser/src/lexer.rs

```diff
@@ -175,8 +175,8 @@ pub struct Lexer<T: Iterator<Item = char>> {
     pending: Vec<Spanned>,
     // The current location.
     location: TextSize,
-    // The last emitted token.
-    last_emitted: Option<Tok>,
+    // Is the last token an equal sign?
+    last_token_is_equal: bool,
```

Comment:
For the future: Lexing the following as magics may make good error recovery more challenging in the future. But I don't think we need to be concerned with this today because lalrpop doesn't give us good error recovery anyway. Furthermore, recovery requires that the parser informs the lexer about contextual information like this to ensure they make the same assumptions.

```python
a = %

a == % 4

def a(b = %magic): ... # or is this valid?
```

But would it make sense for `lex_and_emit_magic_command` to be stricter in an expression context and require that an identifier comes after?

Is the following a valid magic command:

```python
# We may intentionally decide not to support this
a = (
	%magic
)
```

I'm not saying that we should support it. I'm just curious (because it then gets complicated because you also need to deal with comments and such)

Reply:
Although I'm thinking of using the `Name` token inside the `MagicCommand` token which would be the magic command name. This would be tokenized as an identifier right after the magic prefix. I was wondering if we should raise a lexer error if there are any whitespaces between the prefix and the name. I'm wondering because IPython parses it correctly although it's not actually a valid command, but then again this is the job of a linter so we might want to skip the whitespaces.

```python
t.transform_cell("% foo")
# "get_ipython().run_line_magic('', 'foo')\n"

# % foo
# UsageError: Line magic function `%` not found.

t.transform_cell("%foo")
# "get_ipython().run_line_magic('foo', '')\n"
```

Reply:
I don't think the lexer should warn about it. There are two options in the lexer:

* Only lex a magic command if the next character is the start of an identifier. The result is that `% foo` results in an unrecognized token which the parser will handle 
* Lex two distinct tokens (`%` and `Name`) and check the presence of whitespace inside of the Parser (not possible with lalrpop but you could do this in a hand written parser). 

I think I would prefer option 1 for now.

Reply:
Yes, I would prefer option 1 as well. A few things to node:
* There's no `?` token, so in that case it'll raise a lexical error while in other cases it'll go to the parser. This creates an inconsistency w.r.t lexing the same token.
* To avoid the infinite loop of parsing the same character over and over in case of a space, the function `lex_and_emit_magic_command` will have to signal the caller that "Hey, I'm not going to lex this as a magic command" using a boolean or `Result`.

Reply:
(opened #39)